README.md
# <a name="title"></a> Busser::RunnerPlugin::Serverspeclegacy

[![Gem Version](https://badge.fury.io/rb/busser-serverspeclegacy.svg)](http://rubygems.org/gems/busser-serverspeclegacy) [![Build Status](https://secure.travis-ci.org/test-kitchen/busser-serverspeclegacy.svg?branch=master)](https://travis-ci.org/test-kitchen/busser-serverspeclegacy)

A Busser runner plugin for Serverspec
This is a FORK for copatinibility with Chef 12 and Ruby 2.3. Rename you `test/integration/*/serverspec` to `test/integration/*/serverspeclegacy` to use it.

```bash
ls -d1 test/integration/*/serverspec | xargs -I{} mv {}{,legacy}
```

The only difference with original gem is pinned `install_gem('bundler', '~> 1.16.1')` in postinstall, credits to [etherops](https://github.com/etherops), see the [comment](https://github.com/test-kitchen/busser-serverspec/issues/55#issuecomment-468440943)

## Status

This software project is no longer under active development as it has no active maintainers. The software may continue to work for some or all use cases, but issues filed in GitHub will most likely not be triaged. If a new maintainer is interested in working on this project please come chat with us in #test-kitchen on Chef Community Slack.

## <a name="installation"></a> Installation and Setup

Please read the Busser [plugin usage][plugin_usage] page for more details.

## <a name="usage"></a> Usage

Please put test files into [COOKBOOK]/test/integration/[SUITES]/serverspeclegacy/

```cookbook
`-- test
    `-- integration
        `-- default
            `-- serverspeclegacy
                |-- Gemfile
                |-- localhost
                |   `-- httpd_spec.rb
                `-- spec_helper.rb
```

`Gemfile` is optional. You can specify installing Serverspec version and install gems you need.

## <a name="note"></a> Note

### <a name="spec"></a> File Matching

Globbing pattern to match files is `"serverspeclegacy/*/*_spec.rb"`.
You need to use `"_spec.rb"` (underscore), not `"-spec.rb"` (minus).

### <a name="serverspec1"></a> Specify Serverspec version

If you have to specify Serverspec version, you can use Gemfile. Example Gemfile:

```Gemfile
source 'https://rubygems.org'
gem 'serverspec', '< 2.0'
```

### <a name="backend"></a> Serverspec backend

It runs on a target server for testing after ssh log in it.
So you need to specify `set :backend, :exec` not `set :backend, :ssh` (Serverspec v2).
If you use Serverspec v1, you need to specify `include SpecInfra::Helper::Exec` not `include SpecInfra::Helper::Ssh`.


## <a name="development"></a> Development

* Source hosted at [GitHub][repo]
* Report issues/questions/feature requests on [GitHub Issues][issues]

Pull requests are very welcome! Make sure your patches are well tested.
Ideally create a topic branch for every separate change you make. For
example:

1. Fork the repo
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## <a name="authors"></a> Authors

Created and maintained by [HIGUCHI Daisuke][author] (<d-higuchi@creationline.com>)

## <a name="license"></a> License

Apache 2.0 (see [LICENSE][license])


[author]:           https://github.com/cl-lab-k
[issues]:           https://github.com/test-kitchen/busser-serverspec/issues
[license]:          https://github.com/test-kitchen/busser-serverspec/blob/master/LICENSE
[repo]:             https://github.com/test-kitchen/busser-serverspec
[plugin_usage]:     http://docs.kitchen-ci.org/busser/plugin-usage


CHANGELOG.md
# You should use Kitchen::Verifier::Shell + Serverspec

## 0.5.13 2023-02-03

The only difference is pinned `install_gem('net-ssh', '< 7.0')` in postinstall cuz I'm lazy to add it into all cookbooks.

## 0.5.12 2023-01-13

FORK
Version for Ruby 2.3 (Chef 12)
The only difference is pinned `install_gem('bundler', '~> 1.16.1')` in postinstall.

## 0.5.10 2016-08-12

* use container-based test at travis-ci.
* Make rake and rspec/core runtime deps
  merge pull-req by John Keiser <john@johnkeiser.com>, thanks.
* Pin versions of aruba and cucumber
  inspired by test-kitchen/busser-cucumber@cd3e6b3 by Peter Gallagher, thanks.

## 0.5.9 2016-01-22

* Fix frozen string issue when checking Ruby version
  merge pull-req by Joe Friedl <jfriedl@digitalocean.com>, thanks.

## 0.5.8 / 2016-01-22, 0.6.0 / 2016-01-18 (yanked 2016-02-22)

* Install net-ssh < 2.10 for ruby < 2.0.
  merge pull-req by Bill Ruddock <bill.ruddock@gmail.com>, thanks.

## 0.5.7 / 2015-06-01

* use RbConfig instead of obsolete and deprecated Config, closes #29
  suggested by Jörg Herzinger <joerg.herzinger@oiml.at>, thanks.

## 0.5.6 / 2015-04-27

* Run bundle with explicit path to ruby bin, closes #27
  merge pull-req by Simon Detheridge <simon@widgit.com>, thanks.

## 0.5.5 / 2015-04-02

* use File::PATH_SEPARATOR to join ENV PATH. closes #24
  suggested by Jacob McCann <jmccann.git@gmail.com>, thanks.

## 0.5.4 / 2015-03-30

* add the config bindir as well.
  merge pull-req by Noah Kantrowitz <noah@coderanger.net>. thanks.

## 0.5.3 / 2014-10-17

* make private working methods, closes #18

## 0.5.2 / 2014-10-16

* reset gems db

## 0.5.1 / 2014-10-16

* move Gemfile process from postinstall to test.

## 0.5.0 / 2014-10-16

* enable to specify serverspec version in Gemfile, closes #16
* add 'log_switch', '~> 0.3.0', ref: turboladen/tailor#160
* drop ruby 1.9.2 test, because mime-types 2.4.2 does not support ruby 1.9.2.

## 0.4.0 / 2014-10-06

* add --default-path for RSpec 3, closes #15
* bump up version number (0.3: load Gemfile, 0.4: serverspec v2).

## 0.2.8 / 2014-10-05

* update for serverspec v2, closes #13

## 0.2.7 / 2014-09-24

* support loading Gemfile. closes #12

## 0.2.6 / 2014-01-25

* add globbing pattern note for frequency asked test failure.
* update links for moving to test-kitchen organization.
* update copyright.

## 0.2.5 / 2013-11-14

* issue #5: Added --color and --format documentation rspec options
  merge pull-req by Kannan Manickam <kannan@rightscale.com>. thanks.

## 0.2.4 / 2013-11-01

* issue #4: Search all GEM_PATH values for `rspec` bin.
  merge pull-req by Seth Chisamore <schisamo@opscode.com>. thanks.

## 0.2.3 / 2013-07-10

* issue #2: remove tight dependency of serverspec. thanks to rteabeault.

## 0.2.2 / 2013-07-09

* issue #2: remove dependency rspec.

## 0.2.1 / 2013-06-18

* fix silly exit value.

## 0.2.0 / 2013-06-18

* issue #1: do not print full backtrace if serverspec test failure.

## 0.1.1 / 2013-06-06

* add cucumber test and fix related things

## 0.1.0 / 2013-05-22

* Initial release


lib/busser/serverspeclegacy/version.rb
# -*- encoding: utf-8 -*-
#
# Author:: HIGUCHI Daisuke (<d-higuchi@creationline.com>)
#
# Copyright (C) 2013-2014, HIGUCHI Daisuke
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

module Busser

  module Serverspeclegacy

    # Version string for the Serverspec Busser runner plugin
    VERSION = '0.5.13'
  end
end


lib/busser/runner_plugin/serverspeclegacy.rb
# -*- encoding: utf-8 -*-
#
# Author:: HIGUCHI Daisuke (<d-higuchi@creationline.com>)
#
# Copyright (C) 2013-2014, HIGUCHI Daisuke
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'busser/runner_plugin'
require 'rubygems/dependency_installer'

# A Busser runner plugin for Serverspec.
#
# @author HIGUCHI Daisuke <d-higuchi@creationline.com>
#
class Busser::RunnerPlugin::Serverspeclegacy < Busser::RunnerPlugin::Base
  postinstall do
    install_gem('bundler', '~> 1.16.1')
    install_gem('net-ssh', '< 7.0')
  end

  def test
    run_bundle_install
    install_serverspec

    runner = File.join(File.dirname(__FILE__), %w{.. serverspeclegacy runner.rb})
    run_ruby_script!("#{runner} #{suite_path('serverspeclegacy').to_s}")
  end

  private

  def run_bundle_install
    # Referred from busser-shindo
    gemfile_path = File.join(suite_path, 'serverspeclegacy', 'Gemfile')
    if File.exists?(gemfile_path)
      # Bundle install local completes quickly if the gems are already found
      # locally it fails if it needs to talk to the internet. The || below is
      # the fallback to the internet-enabled version. It's a speed optimization.
      banner('Bundle Installing..')
      ENV['PATH'] = [ENV['PATH'], Gem.bindir, RbConfig::CONFIG['bindir']].join(File::PATH_SEPARATOR)
      bundle_exec = "#{File.join(RbConfig::CONFIG['bindir'], 'ruby')} " +
        "#{File.join(Gem.bindir, 'bundle')} install --gemfile #{gemfile_path}"
      run("#{bundle_exec} --local || #{bundle_exec}")
    end
  end

  def install_serverspec
    Gem::Specification.reset
    if Array(Gem::Specification.find_all_by_name('serverspec')).size == 0
      if Gem::Version.new(RUBY_VERSION.dup) < Gem::Version.new('2.3')
        raise "Ruby < 2.3 is EOLed and no longer supported"
      end
      banner('Installing Serverspec..')
      spec = install_gem('serverspec')
      banner "serverspec installed (version #{spec.version})"
    end
  end
end
